scripts/globals/weaponskills.lua

```diff
@@ -16,10 +16,317 @@ require("scripts/globals/magic")
+
+    local levelCorrection = 0
+    if attacker:getMainLvl() < defender:getMainLvl() then
+        levelCorrection = 0.025 * (defender:getMainLvl() - attacker:getMainLvl())
+    end
+
+    cratio = cratio - levelCorrection
+    cratio = cratio * atkmulti
+
+    if cratio > 3 - levelCorrection then
+        cratio = 3 - levelCorrection
+    end
+
+    if cratio < 0 then
+        cratio = 0
+    end
+
+    -- max
+    local pdifmax = 0
+
+    if cratio < 0.9 then
+        pdifmax = cratio * (10 / 9)
+    elseif cratio < 1.1 then
+        pdifmax = 1
+    else
+        pdifmax = cratio
+    end
+
+    -- min
+    local pdifmin = 0
+
+    if cratio < 0.9 then
+        pdifmin = cratio
+    elseif cratio < 1.1 then
+        pdifmin = 1
+    else
+        pdifmin = (cratio * (20 / 19)) - (3 / 19)
+    end
+
+    local pdif = {}
+    pdif[1] = pdifmin
+    pdif[2] = pdifmax
+    -- printf("ratio: %f min: %f max %f\n", cratio, pdifmin, pdifmax)
+    local pdifcrit = {}
+
+    pdifmin = pdifmin * 1.25
+    pdifmax = pdifmax * 1.25
+
+    pdifcrit[1] = pdifmin
+    pdifcrit[2] = pdifmax
+
+    return pdif, pdifcrit
+end
+
+local function getRangedHitRate(attacker, target, capHitRate, bonus)
+    local acc = attacker:getRACC()
+    local eva = target:getEVA()
+
+    if bonus == nil then
+        bonus = 0
+    end
+
+    if (target:hasStatusEffect(xi.effect.YONIN) and target:isFacing(attacker, 23)) then -- Yonin evasion boost if defender is facing attacker
+        bonus = bonus - target:getStatusEffect(xi.effect.YONIN):getPower()
+    end
+
+    if attacker:hasTrait(76) and attacker:isBehind(target, 23) then --TRAIT_AMBUSH
+        bonus = bonus + attacker:getMerit(xi.merit.AMBUSH)
+    end
+
+    acc = acc + bonus
+
+    if attacker:getMainLvl() > target:getMainLvl() then -- acc bonus!
+        acc = acc + ((attacker:getMainLvl() - target:getMainLvl()) * 4)
+    elseif attacker:getMainLvl() < target:getMainLvl() then -- acc penalty :(
+        acc = acc - ((target:getMainLvl() - attacker:getMainLvl()) * 4)
+    end
+
+    local hitdiff = 0
+    local hitrate = 75
+
+    if acc > eva then
+        hitdiff = (acc - eva) / 2
+    end
+
+    if eva > acc then
+        hitdiff = ((-1) * (eva - acc)) / 2
+    end
+
+    hitrate = hitrate + hitdiff
+    hitrate = hitrate / 100
+
+    -- Applying hitrate caps
+    if capHitRate then -- this isn't capped for when acc varies with tp, as more penalties are due
+        if hitrate > 0.95 then
+            hitrate = 0.95
+        end
+
+        if hitrate < 0.2 then
+            hitrate = 0.2
+        end
+    end
+
+    return hitrate
+end
+
 -- Function to calculate if a hit in a WS misses, criticals, and the respective damage done
-function getSingleHitDamage(attacker, target, dmg, wsParams, calcParams)
+local function getSingleHitDamage(attacker, target, dmg, wsParams, calcParams)
     local criticalHit = false
-    local pdif = 0
```

Comment:
An unconventional request on my part, but could you comment this one instead of deleting it?
Mainly as a reminder.